main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala

```diff
@@ -156,29 +159,64 @@ object TypeReduction {
     val candidateMethods = clazz.getMethods
       .filter(m => m.getName == method)
       .filter(m => m.getParameterCount == ts.length)
-      .filter(m =>
-        (m.getParameterTypes zip ts).foldLeft(true){ case (acc, (clazz, tpe)) => acc && (Type.getFlixType(clazz) == tpe) } // Parameter types correspondance
-      )
+      .filter(m => // Parameter types correspondance with subtyping
+        (m.getParameterTypes zip ts).foldLeft(true) {
+          case (acc, (clazz, tpe)) => acc && isSubtype(Type.getFlixType(clazz), tpe)
+          })
+      // NB: once methods with same signatures have been filtered out, we should remove super-methods duplicates if superclass is abstract or ignore if it is a primitive type, e.g., void
+      .filter(m => m.getReturnType.isPrimitive || !java.lang.reflect.Modifier.isAbstract(m.getReturnType.getModifiers)) // temporary to avoid superclass abstract duplicate
 
     candidateMethods.length match {
+      case 0 => JavaResolutionResult.MethodNotFound
       case 1 =>
         val tpe = Type.Cst(TypeConstructor.JvmMethod(candidateMethods.head), loc)
         JavaResolutionResult.Resolved(tpe)
-      case _ => JavaResolutionResult.MethodNotFound // For now, method not found either if there is an ambiguity or no method found
+      case _ =>
+        // Among candidate methods if there already exists the method with the exact signature, we should ignore the rest. E.g.: append(String), append(Object) in SB for the call append("a") should already know to use append(String)
+        // TODO: there is surely a better way to repeat the operation, pre-factorization should be possible
+        val exactSignature = candidateMethods
+          .filter(m => // Parameter types correspondance with subtyping
+            (m.getParameterTypes zip ts).foldLeft(true) {
+              case (acc, (clazz, tpe)) => acc && (Type.getFlixType(clazz) == tpe)
+            })
+        exactSignature.length match {
+          
+          
+          case 1 => JavaResolutionResult.Resolved(Type.Cst(TypeConstructor.JvmMethod(exactSignature.head), loc))
+          case _ => JavaResolutionResult.AmbiguousMethod(candidateMethods.toList)
+        }
+    }
+  }
+
+  /**
+   * Helper method to define a sub-typing relation between two given Flix types.
+   * Returns true if tpe2 is a sub-type of type tpe1, false otherwise.
+   */
+  private def isSubtype(tpe1: Type, tpe2: Type)(implicit flix: Flix): Boolean = {
```

Comment:
The order is backward for my intuition. In Scala, I would expect to read this as "tpe1 is a subtype of tpe2".

Reply:
You should search the codebase for an `isSubtype function. I think we already have one somewhere.

Reply:
I did search a bit and didn’t find anything for Type, I’ll look again later. Magnus expected me to do a new helper method of subtyping.

Reply:
We may have such a function, but I think we will need a new function here anyway.